File: Cloud Storage Samples/OneDrive Browser/cpp/onedrive.cpp
/*
 * Cloud Storage 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/cloudstorage.h"
#define LINE_LEN 100

class MyOneDrive : public OneDrive
{
public:
	
	virtual int FireSSLServerAuthentication(OneDriveSSLServerAuthenticationEventParams *e)
	{
		e->Accept = true;
		return 0;
	}
};

MyOneDrive onedrive1;

void printOptions()
{
	printf( "Type ? to select this menu, a number 1-5 to select the following choices:\n"
	        "1.  List Documents\n"
	        "2.  Download Document\n"
	        "3.  Upload Document\n"
	        "4.  Delete Document\n"
	        "5.  Quit\n");
}

char* resolveResourceType(int type) {
	switch (type)
	{
		case ODRT_FILE: return "File";
		case ODRT_FOLDER: return "Folder";
		default: return "Unknown";
	}
}

void displayDocuments(){
	onedrive1.SetRemoteId(""); // Clear
	onedrive1.SetRemotePath("");
	int ret_code = onedrive1.ListResources();
	if(ret_code)
		goto error;

	printf("%-3.1s%-40.38s%-10.8s%-25.23s\n\n","#","Title","Type","Modified Date");

	for(int i=0;i<onedrive1.GetResourceCount();i++)
	{
		printf("%-3.1d%-40.38s%-10.8s%-25.23s\n", i, onedrive1.GetResourceName(i), resolveResourceType(onedrive1.GetResourceType(i)), onedrive1.GetResourceModifiedTime(i));
	}
	return;

error:
	printf("\nError: %d", ret_code);
	if (onedrive1.GetLastError())
	{
		printf( " \"%s\"\n", onedrive1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	exit(ret_code);
}

int main(int argc, char **argv)
{	
  char buffer[LINE_LEN+1];
  int ret_code = 0;

  printf("***************************************************************\n");
  printf("* This demo shows how to use the OneDrive component to list,  *\n");
  printf("* download, upload, and delete documents.                     *\n");
  printf("***************************************************************\n\n");

  printf("Client ID: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  onedrive1.SetOAuthClientId(buffer);
  
  printf("Client Secret: ");
  fgets(buffer, LINE_LEN, stdin);
  buffer[strlen(buffer)-1] = '\0';
  onedrive1.SetOAuthClientSecret(buffer);

  onedrive1.SetOAuthServerAuthURL("https://login.microsoftonline.com/common/oauth2/v2.0/authorize");
  onedrive1.SetOAuthServerTokenURL("https://login.microsoftonline.com/common/oauth2/v2.0/token");
  onedrive1.Config("OAuthWebServerPort=7777"); // http://localhost:7777 is a registered redirect_url for the app
  onedrive1.SetOAuthAuthorizationScope("offline_access files.readwrite files.readwrite.all");

  printf("\nPerforming OAuth authentication ...\n");

  onedrive1.Authorize();

  ret_code = onedrive1.GetLastErrorCode();

  if(ret_code)
    goto error;

	printf("\n");
	printOptions();

	while (1)
	{
		printf( "\n> " );
		scanf("%80s",buffer);

		if ( ! strcmp(buffer, "?") )
		{
			printOptions();
		}

		else if ( ! strcmp(buffer, "1") ) //List documents
		{
			displayDocuments();
		}

		else if ( ! strcmp(buffer, "2") ) //Download document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);
			int docNum = atoi(buffer);
			
			printf("Download Directory: ");
			scanf("%80s",buffer);
			strcat(buffer,"\\");
			strcat(buffer,onedrive1.GetResourceName(docNum));
			onedrive1.SetLocalFile(buffer);
			onedrive1.SetRemotePath(onedrive1.GetResourcePath(docNum));

			ret_code = onedrive1.DownloadFile(); //Use the default file format
			if(ret_code)
				goto error;

			printf("\n Download Successful\n");
		}

		else if ( ! strcmp(buffer, "3") ) //Upload document
		{
			printf("\nLocal File: ");
			scanf("%80s",buffer);
			onedrive1.SetLocalFile(buffer);
			
			printf("Name of New File on Server: ");
			scanf("%80s",buffer);
            onedrive1.UploadFile(buffer);
			ret_code = onedrive1.GetLastErrorCode();

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "4") ) //Delete Document
		{
			printf("\nDocument Number: ");
			scanf("%80s",buffer);
			int docNum = atoi(buffer);

			onedrive1.SetRemoteId(onedrive1.GetResourceId(docNum));

			onedrive1.DeleteResource();
			ret_code = onedrive1.GetLastErrorCode();

			if(ret_code)
				goto error;

			displayDocuments();
		}

		else if ( ! strcmp(buffer, "5") )
		{
			exit(0);
		}

		else if ( ! strcmp(buffer, "") )
		{
			// Do nothing
		}
		else
		{
			printf( "Please select a number from [1-5].\n" );
		} // end of command checking
	}  // end of main while loop
	return ret_code;

error:
	printf("\nError: %d", ret_code);
	if (onedrive1.GetLastErrorCode())
	{
		printf( " \"%s\"\n", onedrive1.GetLastError() );
	}
	printf("Press any key to continue");
	fflush(stdin);
	getchar();
	return ret_code;

}

File: Cloud Storage Samples/OneDrive Browser/java/onedrive.java
/*
 * Cloud Storage 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Storage in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudstorage
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import cloudstorage.*;

public class onedrive extends ConsoleDemo {
    Onedrive onedrive1 = new Onedrive();

    public onedrive() {
        String buffer = "";

        try {
            onedrive1.addOnedriveEventListener(new cloudstorage.DefaultOnedriveEventListener() {
                public void SSLServerAuthentication(
                        OnedriveSSLServerAuthenticationEvent e) {
                    e.accept = true;
                }
            });

            onedrive1.getOAuth().setClientId(prompt("OAuth Client Id", ":"));
            onedrive1.getOAuth().setClientSecret(prompt("OAuth Client Secret", ":"));
            onedrive1.getOAuth().setServerAuthURL("https://login.microsoftonline.com/common/oauth2/v2.0/authorize");
            onedrive1.getOAuth().setServerTokenURL("https://login.microsoftonline.com/common/oauth2/v2.0/token");
            onedrive1.config("OAuthWebServerPort=7777"); // http://localhost:7777 is a registered redirect_url for the app
            onedrive1.getOAuth().setAuthorizationScope("offline_access files.readwrite files.readwrite.all");
            onedrive1.authorize();

            printMenu();
            while (!buffer.equals("q")) {
                System.out.print("Enter command: ");
                buffer = input();
                if (buffer.toLowerCase().equals("ls")) {
                    ListDocuments();
                } else if (buffer.toLowerCase().equals("get")) {
                    System.out.print("Document #: ");
                    OneDriveResource r = onedrive1.getResources().get(Integer.valueOf(input()));
                    System.out.print("Local Directory: ");

                    onedrive1.setLocalFile(input() + "\\" + r.getName());
                    onedrive1.setRemoteId(r.getId());
                    onedrive1.downloadFile(); //Use the default file format

                    System.out.println("\nDownload Complete.\n");
                } else if (buffer.toLowerCase().equals("put")) {
                    System.out.print("Local File: ");
                    onedrive1.setLocalFile(input());
                    onedrive1.setRemoteId("");
                    File mFile = new File(onedrive1.getLocalFile());
                    onedrive1.uploadFile(mFile.getName());

                    ListDocuments();
                } else if (buffer.toLowerCase().equals("del")) {
                    System.out.print("Document #: ");
                    OneDriveResource r = onedrive1.getResources().get(Integer.valueOf(input()));

                    onedrive1.setRemoteId(r.getId());
                    onedrive1.deleteResource();

                    ListDocuments();
                } else if (buffer.toLowerCase().equals("q")) {
                    System.exit(0);
                } else if (buffer.toLowerCase().equals("?")) {
                    printMenu();
                }
            }
        } catch (CloudStorageException e) {
            System.out.println(e.getMessage());
            System.exit(e.getCode());
            return;
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }

    }

    private void printMenu() {
        System.out.println("\n?\t-\tHelp\n" + "ls\t-\tList Documents\n"
                + "get\t-\tDownload Document\n" + "put\t-\tUpload Document\n"
                + "del\t-\tDelete Document\n" + "q\t-\tQuit\n");
    }

    private void ListDocuments() {
        try {
            onedrive1.setRemoteId(""); // Clear the documents collection
            onedrive1.setRemotePath("");
            onedrive1.listResources();

            System.out.print("\n");
            System.out.format("%1$-3s%2$-20s%3$-30s%4$-40s\n","#","Title","Type","Last Modified");
            System.out.println("--------------------------------------------------------------------------------");

            for (int i = 0; i < onedrive1.getResources().size(); i++) {
                System.out.format("%1$-3s%2$-20.18s%3$-30s%4$-40s\n", String.valueOf(i), onedrive1.getResources().get(i).getName(),
                    ResolveResourceType(onedrive1.getResources().get(i).getType()), onedrive1.getResources().get(i).getModifiedTime());
            }

            System.out.print("\n");

        } catch (CloudStorageException e) {
            System.out.println(e.getMessage());
            System.exit(e.getCode());
            return;
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
    }

    private String ResolveResourceType(int type) {
        switch (type) {
            case OneDriveResource.odrtFile:	return "File";
            case OneDriveResource.odrtFolder: return "Folder";
            default: return "Unknown";
        }
    }

    public static void main(String[] args) {
        try {
            new onedrive();
        } catch (Exception ex) {
            System.out.println("Exception: " + ex.getMessage());
        }
    }

}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudStorageException) {
      System.out.print(" (" + ((CloudStorageException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}



